### PR TITLE
AP_UAVCAN ESC Status Message [from OlliW]

### PR DESCRIPTION
**Pull request by auturgy**
see https://github.com/ArduPilot/ardupilot/issues/7753
Thanks to @olliw42

### CODE REVIEW

File: libraries/AP_UAVCAN/AP_UAVCAN.cpp [hunk 1]
```diff
@@ -1116,4 +1160,51 @@ void AP_UAVCAN::update_mag_state(uint8_t node, uint8_t sensor_id)
     }
 }
 
+AP_UAVCAN::EscStatus_Data* AP_UAVCAN::escstatus_getptrto_data(uint8_t id)
+{
+    // I think the esc_index are continues, by how ArduPilot works
```
Comment (EShamaev) [inline libraries/AP_UAVCAN/AP_UAVCAN.cpp]: "I, me, he, her" please rephrase.
Also please find out if we can jump or not, so on...

File: libraries/AP_UAVCAN/AP_UAVCAN.cpp [hunk 2]
```diff
@@ -454,6 +488,16 @@ bool AP_UAVCAN::try_init(void)
                     esc_raw[_uavcan_i]->setTxTimeout(uavcan::MonotonicDuration::fromMSec(20));
                     esc_raw[_uavcan_i]->setPriority(uavcan::TransferPriority::OneLowerThanHighest);
 
+                    {
```
Comment (EShamaev) [inline libraries/AP_UAVCAN/AP_UAVCAN.cpp]: Why these {} are needed?

File: libraries/AP_UAVCAN/AP_UAVCAN.cpp [hunk 3]
```diff
@@ -1116,4 +1160,51 @@ void AP_UAVCAN::update_mag_state(uint8_t node, uint8_t sensor_id)
...
+    }
+
+    return nullptr;
+}
+
+void AP_UAVCAN::escstatus_update_data(uint8_t id)
+{
+    // only 8 LOG_ESC1_MSG are defined, see /libraries/DataFlash/LogStructure.h
+    // technically, it could happen that the esc_index is not continues, and one would need a better handling
+    // however, I think, ArduPilot implicitly enforces continues esc_index, so should be no problem
```
Comment (EShamaev) [inline libraries/AP_UAVCAN/AP_UAVCAN.cpp]: Please find out exactly.

File: libraries/AP_UAVCAN/AP_UAVCAN.cpp [hunk 4]
```diff
@@ -1116,4 +1160,51 @@ void AP_UAVCAN::update_mag_state(uint8_t node, uint8_t sensor_id)
...
+    if (id >= 8) return;
+
+    for (uint8_t i = 0; i < AP_UAVCAN_ESCSTATUS_MAX_NUMBER; i++) {
+        if (_escstatus.id[i] == id) {
+
```
Comment (EShamaev) [inline libraries/AP_UAVCAN/AP_UAVCAN.cpp]: white space, indenation

**Comment by olliw42** [comment]
many thx, sir
I'm kind of baffled

one comment, as regards using the dataflash singleton: 
In many places in the code I find singletons used with a protection like `s = singleton::instance(); if (s != nullptr) { use singleton; }`, while in many places it is just used as in the code in the above. I really couldn't figure out the pattern when to use what. So, what's currently in the code might be not totally foolproof.

**Comment by olliw42** [comment]
> Why these {} are needed?

I did it that way to make it easier to add support for a new message by copy and pasting. This way one doesn't have to rename the functional irrelevant fields, and I find it also much easier to read. I'm a strong believer in such kind of code things as from my little experience they help significantly to avoid introducing bugs when adding repetitive code. You can note such practice, to make it more visible what's different, in other places too. My habit.
(I think that these repetitive parts eventually should go in a class or so, I however couldn't work that out yet)

**Comment by OXINARF** [comment]
@auturgy Please avoid making branches in our repo, it should always be done in each dev's fork.

**Comment by magicrub** [comment]
I've deleted this branch since it was uploaded to upstream. That force closes the PR. If you still need this branch, I've saved it at magicrub/OlliW_UAVCAN_ESC_FB.

Since it had a merge conflict with recent UAVCAN related merges, I've also rebased the branch and saved it as magicrub/OlliW_UAVCAN_ESC_FB_rebased and opened a new PR https://github.com/ArduPilot/ardupilot/pull/7831

**Pull request closed without merging**